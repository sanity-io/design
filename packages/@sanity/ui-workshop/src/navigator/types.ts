import {WorkshopScope} from '../types'

export interface MenuList {
  type: 'list'
  name?: string
  title?: string
  items: Array<MenuList | MenuScope>
}

export interface MenuScope {
  type: 'scope'
  name: string
  title: string
  scope: WorkshopScope
}

export interface MenuCollection {
  name?: string
  title?: string
  children?: MenuCollection[]
}
