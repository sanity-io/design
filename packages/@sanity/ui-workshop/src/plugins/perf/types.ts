export type PerfTestRunFn<TargetType = unknown> = (context: {
  target: TargetType
}) => Promise<void> | void

export interface PerfTest<TargetType = unknown> {
  description?: string
  name: string
  title?: string
  ref: React.MutableRefObject<TargetType | null>
  run: PerfTestRunFn<TargetType>
}

export interface PerfTestResult {
  name: string
  renders: PerfTestRenderResult[]
  timing?: {
    avgDuration: number
    sumDuration: number
    runs: number
  }
}

export interface PerfTestDetail {
  description?: string
  name: string
  title?: string
}

/**
 * @see React.ProfilerOnRenderCallback
 */
export interface PerfTestRenderResult {
  id: string
  phase: 'mount' | 'update'
  actualDuration: number
  baseDuration: number
  startTime: number
  commitTime: number
  interactions: Set<{
    __count: number
    id: number
    name: string
    timestamp: number
  }>
}

export interface PerfState {
  results: PerfTestResult[]
  activeTest?: string
  testDetails: PerfTestDetail[]
  tests: PerfTest[]
}
