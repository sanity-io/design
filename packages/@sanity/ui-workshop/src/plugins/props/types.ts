export interface GenericPropSchema<T = unknown> {
  name: string
  defaultValue?: T
  groupName?: string
}

export interface BooleanPropSchema extends GenericPropSchema<boolean> {
  type: 'boolean'
}

export interface NumberPropSchema extends GenericPropSchema<number> {
  type: 'number'
}

export type SelectPropValue = string | number | boolean

export type SelectPropOptionsProp<T extends SelectPropValue = SelectPropValue> =
  | Record<PropertyKey, T>
  | Record<Extract<T, PropertyKey>, T[keyof T]>
  | T[]
  | readonly T[]

export interface SelectPropSchema<T extends SelectPropValue = SelectPropValue>
  extends GenericPropSchema<T> {
  type: 'select'
  options: SelectPropOptionsProp<T>
}

export interface StringPropSchema extends GenericPropSchema<string> {
  type: 'string'
}

export interface TextPropSchema extends GenericPropSchema<string> {
  type: 'text'
}

export type PropSchema =
  | BooleanPropSchema
  | NumberPropSchema
  | SelectPropSchema
  | StringPropSchema
  | TextPropSchema

export interface PropsState {
  schemas: PropSchema[]
  value: Record<string, any>
}
