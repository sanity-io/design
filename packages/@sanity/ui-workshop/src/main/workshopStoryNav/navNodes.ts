import {WorkshopScope} from '../../types'
import {MenuCollection} from './types'

export interface NavStoryNode {
  type: 'story'
  name: string
  title?: string
  children: Array<NavScopeNode | NavStoryNode>
}

export interface NavScopeNode {
  type: 'scope'
  name: string
  title?: string
  children: Array<NavScopeNode | NavStoryNode>
}

export type NavNode = NavScopeNode | NavStoryNode

export function buildNavNodes(scopes: WorkshopScope[], collections?: MenuCollection[]): NavNode[] {
  const nodes = []

  for (const scope of scopes) {
    const scopePath = scope.name.split('/')

    nodes.push({
      type: 'scope' as const,
      path: scopePath,
      title: scope.title,
    })

    for (const story of scope.stories) {
      nodes.push({
        type: 'story' as const,
        path: scopePath.concat(story.name.split('/')),
        title: story.title,
        component: story.component,
      })
    }
  }

  const tree: NavNode = {
    type: 'scope',
    name: '$root',
    children: [],
  }

  for (const node of nodes) {
    let parent: NavNode = tree

    for (const name of node.path) {
      let child: NavNode | undefined = parent.children.find((c) => c.name === name)

      if (!child) {
        child = {type: node.type, name, title: node.title, children: []}
        parent.children.push(child!)
      }

      parent = child!
    }
  }

  if (collections) {
    for (const collection of collections) {
      const path = collection.name.split('/')

      const parent = tree

      let child = null

      for (const segment of path) {
        child = parent.children.find((c) => c.name === segment)

        if (!child) {
          break
        }
      }

      if (child) {
        child.title = collection.title
      }
    }
  }

  // console.log('nodes', JSON.stringify(nodes, null, 2))

  // const menu = buildMenu(collections, scopes)

  // console.log('menu', JSON.stringify(menu, null, 2))

  return tree.children
}
