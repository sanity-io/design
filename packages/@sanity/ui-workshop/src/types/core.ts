export interface WorkshopCollection {
  name: string
  title: string
}

export interface WorkshopStory<Options = Record<string, unknown>> {
  name: string
  title: string
  component: React.ComponentType
  options?: Options
}

export interface WorkshopScope {
  name: string
  title: string
  stories: WorkshopStory[]
}
