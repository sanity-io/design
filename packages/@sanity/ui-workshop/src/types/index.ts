export * from './config'
export * from './core'
export * from './location'
export * from './msg'
export * from './plugin'
export * from './state'
