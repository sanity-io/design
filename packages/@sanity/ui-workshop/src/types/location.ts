export type WorkshopQuery = {
  [key: string]: string | number | boolean | WorkshopQuery | undefined
}

export interface WorkshopLocation {
  type: 'pop' | 'push' | 'replace'
  path: string
  query?: WorkshopQuery
}
