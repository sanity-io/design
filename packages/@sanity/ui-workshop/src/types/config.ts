import {WorkshopCollection, WorkshopScope} from './core'
import {WorkshopPlugin} from './plugin'

/** @public */
export interface WorkshopConfig {
  collections?: WorkshopCollection[]
  features?: {
    navbar?: boolean
  }
  frameUrl?: string
  plugins?: WorkshopPlugin[]
  scopes: WorkshopScope[]
  title?: string
}
