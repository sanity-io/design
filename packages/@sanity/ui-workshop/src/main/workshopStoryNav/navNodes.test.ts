import {createElement} from 'react'
import {WorkshopCollection, WorkshopScope} from '../../types'
import {buildNavNodes} from './navNodes'

const component = () => createElement('div')

describe('workshopStoryNav/navNodes', () => {
  it('should build menu', () => {
    const collections: WorkshopCollection[] = [
      {
        name: 'lib1',
        title: 'Library 1',
      },
      {
        name: 'lib2',
        title: 'Library 2',
      },
    ]

    const scopes: WorkshopScope[] = [
      {
        name: 'lib1/first',
        title: 'First scope',
        stories: [
          {
            name: 'story1',
            title: 'Story 1',
            component,
          },
          {
            name: 'story2',
            title: 'Story 3',
            component,
          },
        ],
      },

      {
        name: 'lib2/first',
        title: 'Second scope',
        stories: [
          {
            name: 'story1',
            title: 'Story 1',
            component,
          },
          {
            name: 'story2',
            title: 'Story 3',
            component,
          },
        ],
      },

      {
        name: 'lib2/first/components',
        title: 'Second components',
        stories: [
          {
            name: 'story1',
            title: 'Story 1',
            component,
          },
          {
            name: 'story2',
            title: 'Story 3',
            component,
          },
        ],
      },
    ]

    const nodes = buildNavNodes(scopes, collections)

    expect(nodes).toMatchSnapshot()
  })
})
