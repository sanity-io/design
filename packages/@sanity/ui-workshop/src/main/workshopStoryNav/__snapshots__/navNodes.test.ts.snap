// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workshopStoryNav/navNodes should build menu 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "name": "story1",
            "title": "Story 1",
            "type": "story",
          },
          Object {
            "children": Array [],
            "name": "story2",
            "title": "Story 3",
            "type": "story",
          },
        ],
        "name": "first",
        "title": "First scope",
        "type": "scope",
      },
    ],
    "name": "lib1",
    "title": "Library 1",
    "type": "scope",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "name": "story1",
            "title": "Story 1",
            "type": "story",
          },
          Object {
            "children": Array [],
            "name": "story2",
            "title": "Story 3",
            "type": "story",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "name": "story1",
                "title": "Story 1",
                "type": "story",
              },
              Object {
                "children": Array [],
                "name": "story2",
                "title": "Story 3",
                "type": "story",
              },
            ],
            "name": "components",
            "title": "Second components",
            "type": "scope",
          },
        ],
        "name": "first",
        "title": "Second scope",
        "type": "scope",
      },
    ],
    "name": "lib2",
    "title": "Library 2",
    "type": "scope",
  },
]
`;
